File: Java/src/World/body/Body.java
package World.body;

import GUI.Color_obj;
import World.World;
import World.Point;
import World.body.animal.*;
import World.body.plant.*;

import java.util.concurrent.ThreadLocalRandom;

public abstract class  Body {
    private String name;
    private Color_obj color;
    private int power;
    private int initiative;
    private int age = 0;
    //Value starts form 1
    private Point point_location;
    boolean able_to_action = false;
    boolean alive = true;
    private World world;

    protected void random_location_empty(Body body, Point new_point){
        int x = body.getPoint_location().getX();
        int y = body.getPoint_location().getY();
        int randomNumber;
        boolean tmp = true;
        while(tmp){
            randomNumber = ThreadLocalRandom.current().nextInt(1, 101) % 4 + 1;
            switch (randomNumber){
                case 1:
                    if(x<world.getX_size() && world.getBoxes()[y-1][x].getColor().equals(Color_obj.EMPTY.getColor())){
                        new_point.setX(new_point.getX()+1);
                        tmp=false;
                    }
                    break;
                case 2:
                    if(x > 1 && world.getBoxes()[y-1][x-2].getColor().equals(Color_obj.EMPTY.getColor())){
                        new_point.setX(new_point.getX()-1);
                        tmp = false;
                    }
                    break;
                case 3:
                    if(y<world.getY_size() && world.getBoxes()[y][x-1].getColor().equals(Color_obj.EMPTY.getColor())){
                        new_point.setY(new_point.getY()+1);
                        tmp=false;
                    }
                    break;
                case 4:
                    if(y > 1 && world.getBoxes()[y-2][x-1].getColor().equals(Color_obj.EMPTY.getColor())){
                        new_point.setY(new_point.getY()-1);
                        tmp = false;
                    }
                    break;
            }
        }
    }
    protected  void random_location(Point new_point){
        int x = getPoint_location().getX();
        int y = getPoint_location().getY();
        new_point.setX(x);
        new_point.setY(y);
        int randomNumber;
        boolean tmp = true;
        while(tmp){
            randomNumber = ThreadLocalRandom.current().nextInt(1, 101) % 4 + 1;
            switch (randomNumber){
                case 1:
                    if(x<world.getX_size()){
                        new_point.setX(new_point.getX()+1);
                        tmp=false;
                    }
                    break;
                case 2:
                    if(x > 1){
                        new_point.setX(new_point.getX()-1);
                        tmp = false;
                    }
                    break;
                case 3:
                    if(y<world.getY_size() ){
                        new_point.setY(new_point.getY()+1);
                        tmp=false;
                    }
                    break;
                case 4:
                    if(y > 1){
                        new_point.setY(new_point.getY()-1);
                        tmp = false;
                    }
                    break;
            }
        }
    }

    public void new_body(Point point){
        switch (color){
            case WOLF:
                getWorld().add_body(new Wolf(point, getWorld()));
                break;
            case SHEEP:
                getWorld().add_body(new Sheep(point, getWorld()));
                break;
            case ANTELOPE:
                getWorld().add_body(new Antelope(point, getWorld()));
                break;
            case FOX:
                getWorld().add_body(new Fox(point, getWorld()));
                break;
            case TORTOISE:
                getWorld().add_body(new Tortoise(point, getWorld()));
                break;
            case GRASS:
                getWorld().add_body(new Grass(point, getWorld()));
                break;
            case DANDELION:
                getWorld().add_body(new Dandelion(point, getWorld()));
                break;
            case GUARANA:
                getWorld().add_body(new Guarana(point, getWorld()));
                break;
            case BELLADONNA:
                getWorld().add_body(new Belladonna(point, getWorld()));
                break;
            case SOSNOWSKYSHOGWEED:
                getWorld().add_body(new SosnowskysHogweed(point, getWorld()));
                break;
        }
    }

    public void increment_age(){
        age++;
    }

    public Body(Point point_location, World world) {
        this.point_location = point_location;
        this.world = world;
    }

    public abstract void move();
    protected abstract void collision(Body other);
    protected abstract void action();
    //Getters and setters

    public void setName(String name) {
        this.name = name;
    }

    public void setColor(Color_obj color) {
        this.color = color;
    }

    public void setPower(int power) {
        this.power = power;
    }

    public void setInitiative(int initiative) {
        this.initiative = initiative;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public void setPoint_location(Point point_location) {
        this.point_location = point_location;
    }

    public void setAble_to_action(boolean able_to_action) {
        this.able_to_action = able_to_action;
    }

    public void setAlive(boolean alive) {
        this.alive = alive;
    }

    public void setWorld(World world) {
        this.world = world;
    }

    public String getName() {
        return name;
    }

    public Color_obj getColor() {
        return color;
    }

    public int getPower() {
        return power;
    }

    public int getInitiative() {
        return initiative;
    }

    public int getAge() {
        return age;
    }

    public Point getPoint_location() {
        return point_location;
    }

    public boolean isAble_to_action() {
        return able_to_action;
    }

    public boolean isAlive() {
        return alive;
    }

    public World getWorld() {
        return world;
    }
}

File: Python/World/Body/body.py
import random
from abc import ABC, abstractmethod

from GUI.colors import Colors
from World.point import Point


class Body(ABC):
    def __init__(self, point_location, world):
        self.initiative = None
        self.power = None
        self.name = None
        self.color = None
        self.age = 0
        self.point_location = None
        self.able_to_action = False
        self.alive = True
        self.world = world
        self.point_location = point_location

    def random_location(self, new_point):
        x = self.point_location.get_x()
        y = self.point_location.get_y()
        new_point.set_x(x)
        new_point.set_y(y)
        tmp = True
        while tmp:
            randomNumber = random.randint(1, 4)
            if randomNumber == 1:
                if x < 20:
                    new_point.set_x(new_point.get_x() + 1)
                    tmp = False
            elif randomNumber == 2:
                if x > 1:
                    new_point.set_x(new_point.get_x() - 1)
                    tmp = False
            elif randomNumber == 3:
                if y < 20:
                    new_point.set_y(new_point.get_y() + 1)
                    tmp = False
            elif randomNumber == 4:
                if y > 1:
                    new_point.set_y(new_point.get_y() - 1)
                    tmp = False

    def random_location_empty(self, body, new_point):
        x = body.get_point_location().get_x()
        y = body.get_point_location().get_y()
        # new_point = Point(x, y)
        tmp = True
        while tmp:
            randomNumber = random.randint(1, 4)
            if randomNumber == 1:
                if x < 20 and body.world.boxes[y - 1][x].color.name == 'EMPTY':
                    new_point.set_x(new_point.get_x()+1)
                    tmp = False
            elif randomNumber == 2:
                if x > 1 and body.world.boxes[y - 1][x - 2].color.name == 'EMPTY':
                    new_point.set_x(new_point.get_x()-1)
                    tmp = False
            elif randomNumber == 3:
                if y < 20 and body.world.boxes[y][x - 1].color.name == 'EMPTY':
                    new_point.set_y(new_point.get_y()+1)
                    tmp = False
            elif randomNumber == 4:
                if y > 1 and body.world.boxes[y - 2][x - 1].color.name == 'EMPTY':
                    new_point.set_y(new_point.get_y()-1)
                    tmp = False

    def new_body(self, point):
        from World.Body.Animal.wolf import Wolf
        from World.Body.Animal.antelope import Antelope
        from World.Body.Animal.sheep import Sheep
        from World.Body.Animal.fox import Fox
        from World.Body.Animal.tortoise import Tortoise
        if self.color == Colors.WOLF:
            self.world.add_body(Wolf(point, self.world))
        elif self.color == Colors.SHEEP:
            self.world.add_body(Sheep(point, self.world))
        elif self.color == Colors.ANTELOPE:
            self.world.add_body(Antelope(point, self.world))
        elif self.color == Colors.FOX:
            self.world.add_body(Fox(point, self.world))
        elif self.color == Colors.TORTOISE:
            self.world.add_body(Tortoise(point, self.world))
        # elif self.color == Colors.GRASS:
        #     # self.world.add_body(Grass(point, self.world))
        # elif self.color == Colors.DANDELION:
        #     # self.world.add_body(Dandelion(point, self.world))
        # elif self.color == Colors.GUARANA:
        #     # self.world.add_body(Guarana(point, self.world))
        # elif self.color == Colors.BELLADONNA:
        #     # self.world.add_body(Belladonna(point, self.world))
        # elif self.color == Colors.SOSNOWSKYSHOGWEED:
        #     # self.world.add_body(SosnowskysHogweed(point, self.world))

    def increment_age(self):
        self.age += 1

    @abstractmethod
    def move(self):
        pass

    @abstractmethod
    def collision(self, other):
        pass

    @abstractmethod
    def action(self):
        pass

    # Getters and setters

    def set_name(self, name):
        self.name = name

    def set_power(self, power):
        self.power = power

    def set_initiative(self, initiative):
        self.initiative = initiative

    def set_age(self, age):
        self.age = age

    def set_point_location(self, point_location):
        self.point_location = point_location

    def set_able_to_action(self, able_to_action):
        self.able_to_action = able_to_action

    def set_alive(self, alive):
        self.alive = alive

    def set_world(self, world):
        self.world = world

    def set_color(self, color):
        self.color = color

    def get_name(self):
        return self.name

    def get_power(self):
        return self.power

    def get_initiative(self):
        return self.initiative

    def get_age(self):
        return self.age

    def get_point_location(self):
        return self.point_location

    def is_able_to_action(self):
        return self.able_to_action

    def is_alive(self):
        return self.alive

    def get_world(self):
        return self.world

    def get_color(self):
        return self.color

